fix(deadline): adding version check for staging deadline

yashda opened the pull request:
Adding minimum supported deadline version check while staging
deadline installers. Now we fail fast if the provided version
is not supported by RFDK.


----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

yashda committed:
fix(deadline): adding version check for staging deadline

Adding minimum supported deadline version check while staging
deadline installers. Now we fail fast if the provided version
is not supported by RFDK.

yashda committed:
fix(deadline): adding version check for staging deadline

Adding minimum supported deadline version check while staging
deadline installers. Now we fail fast if the provided version
is not supported by RFDK.

ddneilson committed:
fix(examples): Minor fixes to Python example app REAME

ryyakobe commented on packages/aws-rfdk/lib/core/lib/utils.ts:
```diff
@@ -0,0 +1,31 @@
+   * than second version.
+   */
+  public static versionCompare(versionFirst: string, versionSecond: string): number {
+    const regExStripZero = /(\.0+)+$/;
```
I wonder if we should handle cases with a trailing dot: `10.1.0.`.
Not sure if we need it, but something like `(\.0+.+)+$` will do the trick.

What do you think?

yashda replied:
The method 'versionCompare' assumes that the version strings are already formatted. I added another validation for the version format. So, if the version string itself is incorrect, we will throw an error accordingly. This method will continue to just compare the formatted version strings.

ryyakobe commented on packages/aws-rfdk/lib/core/lib/utils.ts:
```diff
@@ -0,0 +1,31 @@
+   * @param versionSecond
+   *
+   * @returns negative value if first version is smaller than second version;
+   * 0 if the versions matches, positive value if first version is smaller
```
There is an error in the description. These 2 statements contradict:

***negative*** value if first version is smaller than second version
***positive***  value if first version is smaller than second version.
```suggestion
   * 0 if the versions matches, positive value if first version is greater
```

yashda replied:
fixed. good catch.

ryyakobe commented on packages/aws-rfdk/lib/core/lib/utils.ts:
```diff
@@ -0,0 +1,31 @@
+    const l = Math.min(versionArrayFirst.length, versionArraySecond.length);
+
+    for (let i = 0; i < l; i++) {
+      let diff = parseInt(versionArrayFirst[i], 10) - parseInt(versionArraySecond[i], 10);
```
Neat: I guess you should be able to use const here.
```suggestion
      const diff = parseInt(versionArrayFirst[i], 10) - parseInt(versionArraySecond[i], 10);
```

yashda replied:
updated.

ryyakobe commented on packages/aws-rfdk/lib/deadline/lib/stage.ts:
```diff
@@ -154,6 +157,12 @@ export class Stage {
     }
 
+    // Do minimum supported deadline version check
+    const minimumSupportedVersion = '10.1.9';
```
Souldn't we check for `10.1.9.2` here? We state in our documentation that we support this version, but this looks to work with `10.1.9.0`.

yashda replied:
'2' is the only patch that was released under '10.1.9'. So its safe to compare it with '10.1.9' in the code.

yashda committed:
review feedback

yashda committed:
Merge remote-tracking branch 'upstream/mainline' into yashdaStageVersionCheck

yashda committed:
moving method to Utils class

jusiskin commented on packages/aws-rfdk/lib/core/lib/utils.ts:
```diff
@@ -0,0 +1,54 @@
+/**
```
Two suggestions:

1.  Rather than starting a `utils.ts` file, what would you think about renaming this file to `version.ts`? It would be good to avoid the common pitfall of having a growing utils module that becomes a collection of miscellaneous code.
1.  How would you feel about making this object-oriented? I'm thinking a `Version` class would be nice, where instances could be constructed in two ways - either by specifying the numeric components or parsed from a string:

    ```ts
    class Version {
    public readonly components: number[];

    constructor(...components: number[]) {
        this.components = components;
    }

    public isGreaterThan(other: Version): boolean {
        // code here
    }

    public static parse(versionText: string): Version {
        // code to parse a version string
        return new Version(...components);
    }
    }
    ```

Putting this together, it makes the API and internal code a bit more intuitive/readable:

```ts
import { Version } from './version';

const versionA = new Version(10, 1, 9, 2);
const versionB = Version.parse('10.1.9.3');
if (versionB.isGreaterThan(versionA)) {
  console.log('B is greater than A');
}
```

What do you think?

yashda replied:
sounds good. I have modified it accordingly. Also removed a duplicate version parsing logic.

jusiskin commented on packages/aws-rfdk/lib/core/lib/utils.ts:
```diff
@@ -0,0 +1,54 @@
+     * - * Matches previous pattern zero or more times.
+     * - $ asserts position at the end of a line
+     */
+    if (version.match(/^\d+(?:\.\d+)*$/g)) {
```
nit: we don't require the `/g` flag here, since we are anchoring to the beginning (`^`) and end (`$`) of the string.

yashda replied:
fixed.

jusiskin commented on packages/aws-rfdk/lib/core/lib/utils.ts:
```diff
@@ -0,0 +1,54 @@
+        return diff;
+      }
+    }
+    return versionArrayFirst.length - versionArraySecond.length;
```
This function can currently be invoked as:

```ts
const result = versionCompare('10.1', '10.1.9');    // result will be -1
```

Is this the expected behavior? I suppose it depends on the intended semantic meaning when passing `"10.1"` as an argument to this function? Right now it is being treated as an implicit `"10.1.0"`, but I wonder if some users might consider `"10.1"` to mean the latest `10.1.x.y` release.

We cannot resolve versions (`10.1.x.y`) at deploy-time, so perhaps we should throw an exception if the two versions don't have the same number of components when being passed into this function. Callers can add `".0"` if need-be. What do you think?

yashda replied:
Modified the code to only accept full, resolved version string. We can add another method to parse partial, unresoled version string once we add that logic.

jusiskin reviewed (changes requested):
This looks great. My main suggestion is around code organization, but there is another concern about input validation.

ryyakobe commented on packages/aws-rfdk/lib/core/lib/utils.ts:
```diff
@@ -0,0 +1,54 @@
+   * @param versionFirst
+   * @param versionSecond
+   *
+   * @returns negative value if first version is smaller than second version;
```
Neat. Maybe:
```suggestion
   * @returns a negative value if the first version is smaller than the second version;
   * 0 if the versions match;
   * a positive value if the first version is greater than the second version.
```

jusiskin replied:
+1. I'd also suggest replacing "smaller" with "less than"

yashda replied:
updated.

ryyakobe reviewed (commented):
Looks good! Thanks, Yash! Do you think we could change a docstring just a bit more?

horsmand committed:
chore(repo): auto approve Dependabot PR's (#113)

- Adds support for using the auto approve GitHub action to automatically
approve PR's from Dependabot
- Adds support for using Dependabot to update the version of any GitHub
actions we are using

dependabot[bot] committed:
chore(deps): bump typescript from 3.9.7 to 4.0.2 (#107)

Bumps [typescript](https://github.com/Microsoft/TypeScript) from 3.9.7 to 4.0.2.
- [Release notes](https://github.com/Microsoft/TypeScript/releases)
- [Commits](https://github.com/Microsoft/TypeScript/compare/v3.9.7...v4.0.2)

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

aws-painec committed:
chore(integ): rewrite end-to-end test to simplify deploymen

* chore(integ): rewrite end-to-end test to simplify deployment, use fallback values, eliminate output file

* chore(integ): breaks up end-to-end scripts for test components into deploy, test, and teardown stages, use common functions instead of redundant scripts; add fallback to cleanup stacks on test failure

* chore(integ): add summary output of test execution with deployment times

* chore(integ): all tests now assumed to run unless override variable is set

* chore(integ): break end-to-end test into different component scripts

* chore(integ): combinede e2e scripts for test components into one component_e2e.sh script

Co-authored-by: Claire Paine <painec@dev-dsk-painec-2c-9abb0f50.us-west-2.amazon.com>
Co-authored-by: Claire Paine <painec@dev-dsk-painec-2c-041e65bb.us-west-2.amazon.com>

yashda committed:
review feedback

yashda committed:
review feedback

yashda committed:
Merge branch 'mainline' into yashdaStageVersionCheck

ryyakobe commented on packages/aws-rfdk/lib/deadline/lib/stage.ts:
```diff
@@ -154,6 +159,12 @@ export class Stage {
 
+    // Do minimum supported deadline version check
+    const minimumSupportedVersion = new Version([10, 1, 9, 2]);
+    if (minimumSupportedVersion.isGreaterThan(Version.parse(version))) {
```
I feel it's a bit confusing to use a function called `isGreaterThan`, when the message says `is less than`.
You have to reread the code to understand what is really happening. 

Could we have both of these statements match? I guess `<` is used more often than `>`, so I would suggest changing the function to `isLessThan` and use
`if (manifestVersion.isLessThan(minimumSupportedVersion)) {`

What do you think?

ryyakobe commented on packages/aws-rfdk/lib/deadline/lib/stage.ts:
```diff
@@ -154,6 +159,12 @@ export class Stage {
 
+    // Do minimum supported deadline version check
+    const minimumSupportedVersion = new Version([10, 1, 9, 2]);
+    if (minimumSupportedVersion.isGreaterThan(Version.parse(version))) {
```
I feel it's a bit confusing to use a function called `isGreaterThan`, when the message says `is less than`.
You have to reread the code to understand what is really happening. 

Could we have both of these statements match? I guess `<` is used more often than `>`, so I would suggest changing the function to `isLessThan` and use
`if (manifestVersion.isLessThan(minimumSupportedVersion)) {`

What do you think?

yashda replied:
Added another method 'isLessThan' to the Version class.

yashda committed:
adding isLessThan method

ryyakobe reviewed (dismissed):
Great work! Thanks Yash!

jusiskin commented on packages/aws-rfdk/lib/deadline/lib/stage.ts:
```diff
@@ -154,6 +159,13 @@ export class Stage {
     }
 
+    // Do minimum supported deadline version check
+    const minimumSupportedVersion = new Version([10, 1, 9, 2]);
```
#100 introduces a similar constant inline in code with this same minimum Deadline version. Are these minimums linked or logically separate?

Even if separate, I wonder if there's a single place we can specify this value that is easy to find later. This makes it easy if we ever need to change our minimum supported version later. What do you think? I'd also be open to creating a new issue for it for later.

yashda replied:
Agree. I have now modified the code such that the minimum supported deadline version will now be passed as a parameter to the UserData. This makes #100 dependent on this PR.

jusiskin commented on packages/aws-rfdk/lib/deadline/lib/version.ts:
```diff
@@ -72,6 +72,106 @@ abstract class VersionQueryBase extends Construct implements IVersion {
+  public readonly components: number[];
+
+  constructor(components: number[]) {
+    this.components = components;
```
Should we add some input validation here as well and the corresponding tests? We are accepting arbitrary number of components and negative numbers.

yashda replied:
Added.

jusiskin commented on packages/aws-rfdk/lib/deadline/test/version.test.ts:
```diff
@@ -13,11 +13,153 @@ import {
+  });
+
+  describe('.parse', () => {
+    test.each<[string, { version: string, expectedValue: boolean }]>([
```
How would you feel about renaming `expectedValue` to `expectException` to be a bit more specific and readable? The values would need to be inverted.

yashda replied:
Updated.

jusiskin commented on packages/aws-rfdk/lib/deadline/test/version.test.ts:
```diff
@@ -13,11 +13,153 @@ import {
+      } finally {
+      }
+
+      expect(result).toEqual(expectedValue);
```
nit: while this test is technically valid, jest has some nice helpers for asserting exceptions being thrown that cleans up the `try`/`catch` block:

```ts
// Function to call
function parse() {
  Version.parse(version);
}

if (expectException) {
  expect(parse).toThrow();
}
else {
  expect(parse).not.toThrow();
}

```

yashda replied:
updated.

jusiskin commented on packages/aws-rfdk/lib/deadline/lib/version.ts:
```diff
@@ -72,6 +72,106 @@ abstract class VersionQueryBase extends Construct implements IVersion {
+/**
+  * This class is reposonsible to do basic operations on version format.
+  */
+export class Version {
```
It looks like this class could implement the `IPatchVersion` interface here. We could leverage [TypeScript accessors](https://www.typescriptlang.org/docs/handbook/classes.html#accessors) for implementing the individual components, e.g.:

```ts
export class Version {
  // ...
  public get majorVersion(): number {
    return this.components[0];
  }
}
```

This could simplify the code in `VersionQuery.parseVersionString`.

yashda replied:
updated.

jusiskin reviewed (changes requested):
This is looking great. I have a few more suggestions. Let me know what you think.

yashda committed:
review feedback

jusiskin commented on packages/aws-rfdk/lib/deadline/lib/version.ts:
```diff
@@ -116,9 +121,47 @@ export class Version {
   constructor(components: number[]) {
+    // validations
+    if(components.length != 4) {
+      throw new TypeError('Invalid version format. Version should contain exactly 4 components.');
```
Minor suggestion here:

The TypeScript compiler and JSII enforce correct typing here, so rather than throwing a `TypeError`, should we just throw a generic `Error` class for this.

jusiskin commented on packages/aws-rfdk/lib/deadline/lib/version.ts:
```diff
@@ -116,9 +121,47 @@ export class Version {
+    }
+    components.forEach((component) => {
+      if (component < 0) {
+        throw new TypeError('Invalid version format. None of the version components can be negative.');
```
Do you think [`RangeError`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RangeError) fits better here?

jusiskin commented on packages/aws-rfdk/lib/deadline/lib/worker-fleet.ts:
```diff
@@ -548,6 +549,7 @@ export class WorkerInstanceFleet extends WorkerInstanceFleetBase {
         `'${groups}'`,
         `'${pools}'`,
         `'${props.region || ''}'`,
+        `'${Version.MINIMUM_SUPPORTED_DEADLINE_VERSION.toString()}'`,
```
This and the changes in:

*   `packages/aws-rfdk/lib/deadline/scripts/powershell/configureWorker.ps1`
*   `packages/aws-rfdk/lib/deadline/test/worker-fleet.test.ts`

probably belong in #100 since they are not used yet.

jusiskin reviewed (changes requested):
Last few changes - but this is otherwise ready.

yashda committed:
review feedback

jusiskin reviewed (dismissed):
Approved!

ryyakobe commented on packages/aws-rfdk/lib/deadline/lib/version.ts:
```diff
@@ -72,6 +72,149 @@ abstract class VersionQueryBase extends Construct implements IVersion {
+   * @returns true if this version is less than the provided version;
+   * false if this version is greater than or equal to the provided verison.
+   */
+  public isLessThan(version: Version): boolean {
```
There is some code duplication between `isLessThan` and `isGreaterthan`.
I think in some earlier iteration, you already used an integer as return type in `Utils.versionCompare`.
Maybe we can use a something like below to remove the code duplication and to have all 3 comparisons available:

```
private static compare(first: Version, second: Version): number {
  if (first.components.length != second.components.length) {
    throw new TypeError('Component count in both the versions should be same.');
  }

  for (let i = 0; i < first.components.length; i++) {
    const diff = first.components[i] - second.components[i];
    if (diff != 0) {
      return diff;
    }
  }
  return 0;
}


public isGreaterThan(version: Version): boolean {
  return compare(this, version) == 1;
}

public isGreaterThan(version: Version): boolean {
  return compare(this, version) == -1;
}

public isEqual(version: Version): boolean {
  return compare(this, version) == 0;
}
```

yashda replied:
updated

yashda committed:
review feedback

ryyakobe commented on packages/aws-rfdk/lib/deadline/lib/version.ts:
```diff
@@ -72,6 +72,158 @@ abstract class VersionQueryBase extends Construct implements IVersion {
+   * false if this version is less than or equal to the provided verison.
+   */
+  public isGreaterThan(version: Version): boolean {
+    if (this.components.length != version.components.length) {
```
Is there a reason why can't we move this check inside of the `compare` function?

yashda replied:
updated.

ryyakobe commented on packages/aws-rfdk/lib/deadline/lib/version.ts:
```diff
@@ -72,6 +72,158 @@ abstract class VersionQueryBase extends Construct implements IVersion {
+   * 0 if both the versions are equal;
+   * positive value if this version is greater than the provided verison.
+   */
+  private compare(version: Version): number {
```
We still don't have an easy way to compare for version equality.
Could you please add `isEqualTo` method similar to `isGreaterThan` and `isLessThan`?

yashda replied:
added.

yashda committed:
review feedback

ryyakobe reviewed (approved):
Great work! Thanks again, Yash!

jusiskin reviewed (approved):
Approved! Looks great!
